pygatt/bled112_backend.py

```diff
@@ -0,0 +1,1882 @@
+        cmd = self._lib.ble_cmd_gap_connect_direct(bd_addr, addr_type,
+                                                   interval_min, interval_max,
+                                                   timeout, latency)
+        self._lib.send_command(self._ser, cmd)
+
+        # Wait for response
+        self._loglock.release()  # don't hold loglock while waiting
+        self._wait_for_cmd_response()
+        if self._response_return != 0:
+            self._loglock.acquire()
+            self._logger.warn("connect_direct failed: %s",
+                              get_return_message(self._response_return))
+            self._loglock.release()
+            self._main_thread_cond.release()
+            return False
+
+        # Start timeout timer
+        timer.start()
+
+        # Wait for event
+        while not (self._connected or self._connect_timeout):
+            self._main_thread_cond.wait()
+        self._loglock.acquire()
+        if self._connect_timeout:
+            self._connect_timeout = False
+            if not self._connected:
+                self._logger.warn("Connect timeout")
+                self._loglock.release()
+                self._main_thread_cond.release()
+                return False
+
+        # Drop locks
+        self._drop_locks()
+
+        return True
+
+    def delete_stored_bonds(self):
+        """
+        Delete the bonds stored on the BLED112.
+
+        Note: this does not delete the corresponding bond stored on the remote
+              device.
+        """
+        # Get locks
+        self._get_locks()
+
+        # Find bonds
+        self._logger.info("get_bonds")
+        self._stored_bonds = []
+        cmd = self._lib.ble_cmd_sm_get_bonds()
+        self._lib.send_command(self._ser, cmd)
+
+        # Wait for response
+        self._loglock.release()  # don't hold loglock while waiting
+        self._wait_for_cmd_response()
+        if self._num_bonds == 0:  # no bonds
+            self._main_thread_cond.release()
+            return
+
+        # Wait for event
+        while len(self._stored_bonds) < self._num_bonds:
+            self._main_thread_cond.wait()
+
+        # Delete bonds
+        self._loglock.acquire()
+        for b in reversed(self._stored_bonds):
+            self._logger.info("delete_bonding")
+            cmd = self._lib.ble_cmd_sm_delete_bonding(b)
+            self._lib.send_command(self._ser, cmd)
+
+            # Wait for response
+            self._loglock.release()  # don't hold loglock while waiting
+            self._wait_for_cmd_response()
+            self._loglock.acquire()
+            if self._response_return != 0:
+                self._logger.warn("delete_bonding: %s",
+                                  get_return_message(self._response_return))
+                self._loglock.release()
+                self._main_thread_cond.release()
+                return
+
+        # Drop locks
+        self._drop_locks()
+
+    def disconnect(self):
```

Comment:
It is not possible to hang because the BLED112 will respond to the command with an error code. That will cause self._response_return to be non-zero so it will return early.